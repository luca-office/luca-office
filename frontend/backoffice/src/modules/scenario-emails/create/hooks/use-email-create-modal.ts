import isEmpty from "lodash-es/isEmpty"
import {useForm, UseFormMethods} from "react-hook-form"
import {useDispatch} from "react-redux"
import {SelectOptionCustomized} from "shared/components"
import {EmailDirectory, Relevance} from "shared/graphql/generated/globalTypes"
import {useCreateEmail, useScenario, useUpdateScenario} from "shared/graphql/hooks"
import {Email, NavigationConfig, Scenario} from "shared/models"
import {LucaI18nLangKey, useLucaTranslation} from "shared/translations"
import {Option} from "shared/utils"
import {navigateToRouteAction} from "../../../../redux/actions/navigation-action"
import {Route} from "../../../../routes"
import {isOutgoingEmail} from "../../utils/directory"
import {cleanEmailDelay} from "../../utils/timing"

export interface EmailCreationForm {
  readonly directory: EmailDirectory
  readonly receptionDelayInSeconds: string
  readonly senderOrRecipient: string
}

export interface UseEmailCreateModalHook {
  readonly formMethods: UseFormMethods<EmailCreationForm>
  readonly createEmail: (data: EmailCreationForm) => void
  readonly createEmailLoading: boolean
  readonly dismissModal: () => void
  readonly directoryOptions: SelectOptionCustomized[]
  readonly senderOrRecipientInputLabelKey: LucaI18nLangKey
  readonly isConfirmDisabled: boolean
}

export const useEmailCreateModal = (
  scenarioId: UUID,
  navigationConfig: NavigationConfig<Route>,
  isIntroductionEmail: boolean,
  routeToEmailAfterCreation: boolean,
  defaultDirectory: EmailDirectory
): UseEmailCreateModalHook => {
  const {t} = useLucaTranslation()
  const dispatch = useDispatch()

  const {scenario: scenarioOption} = useScenario(scenarioId)
  const {createEmail, createEmailLoading} = useCreateEmail()
  const {updateScenario, isUpdateScenarioLoading} = useUpdateScenario()

  const formMethods = useForm<EmailCreationForm>({
    defaultValues: {directory: defaultDirectory, receptionDelayInSeconds: "0"}
  })

  const directoryOptions = [
    {value: EmailDirectory.Inbox, label: t("email__directory_inbox")},
    {value: EmailDirectory.Sent, label: t("email__directory_sent")},
    {value: EmailDirectory.Trash, label: t("email__directory_trash")}
  ]

  const selectedDirectory = formMethods.watch<keyof Pick<EmailCreationForm, "directory">, EmailDirectory>("directory")
  const senderOrRecipientInputLabelKey = isOutgoingEmail(selectedDirectory)
    ? "email__directory_recipient"
    : "email__directory_sender"

  const isConfirmDisabled = createEmailLoading || isUpdateScenarioLoading || !isEmpty(formMethods.errors)

  const dismissModal = () => {
    dispatch(navigateToRouteAction(navigationConfig.route, navigationConfig.payload))
  }

  const onCreationSuccess = (createdEmail: Option<Email>) =>
    routeToEmailAfterCreation && createdEmail.isDefined()
      ? createdEmail.forEach(mail => {
          dispatch(
            navigateToRouteAction(Route.ScenarioEmails, {scenarioId, emailId: mail.id, directory: mail.directory})
          )
        })
      : dismissModal()

  const checkAndUpdateScenarioMailSettings = (email: Option<Email>) =>
    isIntroductionEmail
      ? email
          .flatMap(({id: emailId, sender}) =>
            scenarioOption.map<Promise<Option<Scenario> | void>>((scenario: Scenario) =>
              updateScenario(scenario.id, {
                date: scenario.date,
                name: scenario.name,
                description: scenario.description,
                maxDurationInSeconds: scenario.maxDurationInSeconds,
                introductionEmailId: emailId,
                shouldDisplayTime: scenario.shouldDisplayTime,
                tags: scenario.tags,
                // LUCA-307 automatically map intro sender to completion sender if not already present
                completionEmailAddress: scenario.completionEmailAddress ? scenario.completionEmailAddress : sender,
                shouldHideReferenceBookChapters: scenario.shouldHideReferenceBookChapters,
                sampleCompanyId: scenario.sampleCompanyId
              })
            )
          )
          .getOrElse(Promise.resolve())
          .then(() => email)
      : email

  const handleEmailCreation = ({senderOrRecipient, directory, receptionDelayInSeconds}: EmailCreationForm) => {
    createEmail({
      sender: isOutgoingEmail(directory) ? null : senderOrRecipient,
      recipient: isOutgoingEmail(directory) ? senderOrRecipient : null,
      ccRecipients: [],
      subject: "",
      directory,
      receptionDelayInSeconds: cleanEmailDelay(parseInt(receptionDelayInSeconds, 10), directory),
      isInitiallyRead: directory !== EmailDirectory.Inbox,
      relevance: Relevance.PotentiallyHelpful,
      message: "",
      scenarioId
    })
      .then(checkAndUpdateScenarioMailSettings)
      .then(onCreationSuccess)
  }

  return {
    formMethods,
    createEmail: handleEmailCreation,
    createEmailLoading: createEmailLoading || isUpdateScenarioLoading,
    dismissModal,
    directoryOptions,
    senderOrRecipientInputLabelKey,
    isConfirmDisabled
  }
}
