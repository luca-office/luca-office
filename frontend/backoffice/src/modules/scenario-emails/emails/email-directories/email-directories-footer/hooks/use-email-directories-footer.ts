import {useDispatch, useSelector} from "react-redux"
import {Payload} from "redux-first-router"
import {Email} from "shared/models"
import {Option} from "shared/utils"
import {navigateToRouteAction} from "../../../../../../redux/actions/navigation-action"
import {AppState} from "../../../../../../redux/state/app-state"
import {Route} from "../../../../../../routes"

export interface UseEmailDirectoriesFooterHook {
  readonly handleCreateClick?: () => void
  readonly showCreateEmailModal?: () => void
}

export const useEmailDirectoriesFooter = (
  actionsDisabled: boolean,
  introductionEmail: Option<Email>
): UseEmailDirectoriesFooterHook => {
  const dispatch = useDispatch()
  const directory = useSelector<AppState, Payload | undefined>(store => store.location?.payload?.directory)
  const emailId = useSelector<AppState, Payload | undefined>(store => store.location?.payload?.emailId)
  const scenarioId = useSelector<AppState, Payload | undefined>(store => store.location?.payload?.scenarioId)

  const showCreateEmailModal = !actionsDisabled
    ? () => {
        if (directory && scenarioId) {
          dispatch(
            navigateToRouteAction(Route.ScenarioEmailsEmailCreation, {
              scenarioId,
              emailId,
              directory
            })
          )
        }
      }
    : undefined

  const handleCreateClick =
    !actionsDisabled || introductionEmail.isDefined()
      ? () => {
          if (introductionEmail.isDefined()) {
            introductionEmail.forEach(({id, directory, scenarioId}) => {
              dispatch(
                navigateToRouteAction(Route.ScenarioEmails, {
                  scenarioId,
                  directory,
                  emailId: id
                })
              )
            })
          } else if (scenarioId && directory) {
            dispatch(
              navigateToRouteAction(Route.ScenarioEmailsIntroductionEmailCreation, {
                scenarioId,
                directory,
                emailId
              })
            )
          }
        }
      : undefined

  return {handleCreateClick, showCreateEmailModal}
}
