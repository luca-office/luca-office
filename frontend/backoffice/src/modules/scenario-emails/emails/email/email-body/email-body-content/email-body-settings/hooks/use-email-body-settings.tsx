import * as React from "react"
import {useDispatch} from "react-redux"
import {SelectOptionCustomized} from "shared/components"
import {NotificationSeverity} from "shared/enums"
import {EmailDirectory, EmailUpdate} from "shared/graphql/generated/globalTypes"
import {AppNotification, Email} from "shared/models"
import {useLucaTranslation} from "shared/translations"
import {Option} from "shared/utils"
import {EmailUnreadIndicator} from "../../../../../../../../components"
import {EmailStatus} from "../../../../../../../../enums"
import {navigateToRouteAction} from "../../../../../../../../redux/actions/navigation-action"
import {updateNotification} from "../../../../../../../../redux/actions/ui/common-ui-action"
import {Route} from "../../../../../../../../routes"
import {getDirectoryOptions} from "../../../../../../utils/directory"

export interface UseEmailBodySettingsHook {
  readonly directoryOptions: SelectOptionCustomized[]
  readonly hideDelayOverlay: () => void
  readonly isDelayOverlayVisible: boolean
  readonly markerOptions: SelectOptionCustomized[]
  readonly moveEmail: (directory: EmailDirectory) => void
  readonly showDelayOverlay: () => void
}

export const useEmailBodySettings = (
  email: Email,
  updateEmail: (update: Partial<EmailUpdate>, onSuccess?: () => void) => Promise<Option<Email>>
): UseEmailBodySettingsHook => {
  const {t} = useLucaTranslation()
  const dispatch = useDispatch()
  const [isDelayOverlayVisible, setDelayOverlayVisible] = React.useState<boolean>(false)
  const switchDirectory = (directory: EmailDirectory) =>
    dispatch(navigateToRouteAction(Route.ScenarioEmails, {scenarioId: email.scenarioId, directory, emailId: email.id}))
  const showMoveWarning = () =>
    dispatch(
      updateNotification(
        Option.of<AppNotification>({
          severity: NotificationSeverity.Warning,
          messageKey: "notification__message_email_folder_delay"
        })
      )
    )
  const moveEmail = (directory: EmailDirectory) => {
    const successFn = () => switchDirectory(directory)

    if (
      email.directory === EmailDirectory.Inbox &&
      email.receptionDelayInSeconds > 0 &&
      directory !== EmailDirectory.Inbox
    ) {
      updateEmail({directory, receptionDelayInSeconds: 0}, () => {
        showMoveWarning()
        successFn()
      })
    } else {
      updateEmail({directory}, successFn)
    }
  }
  const markerOptions = React.useMemo<SelectOptionCustomized[]>(
    () => [
      {
        value: EmailStatus.Read,
        label: t("email__status_read")
      },
      {
        value: EmailStatus.Unread,
        label: t("email__status_unread"),
        renderOptionLabel: label => <EmailUnreadIndicator text={label} />
      }
    ],
    [email]
  )

  const showDelayOverlay = () => setDelayOverlayVisible(true)
  const hideDelayOverlay = () => setDelayOverlayVisible(false)

  return {
    directoryOptions: getDirectoryOptions(t),
    hideDelayOverlay,
    isDelayOverlayVisible,
    markerOptions,
    moveEmail,
    showDelayOverlay
  }
}
