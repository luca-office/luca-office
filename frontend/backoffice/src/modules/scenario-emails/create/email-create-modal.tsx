import {css} from "@emotion/react"
import * as React from "react"
import {Controller} from "react-hook-form"
import {CustomSelect, Modal, Overlay, Text, TextInput} from "shared/components"
import {InputType} from "shared/enums"
import {EmailDirectory} from "shared/graphql/generated/globalTypes"
import {NavigationConfig} from "shared/models"
import {Flex, spacingLarge, spacingMedium} from "shared/styles"
import {useLucaTranslation} from "shared/translations"
import {emailRegexPattern} from "shared/utils"
import {Route} from "../../../routes"
import {useEmailCreateModal} from "./hooks/use-email-create-modal"

export interface CreateEmailModalProps {
  readonly scenarioId: UUID
  readonly navigationConfig: NavigationConfig<Route>
  readonly directory?: EmailDirectory
  readonly isIntroductionEmail?: boolean
  readonly routeToEmailAfterCreation?: boolean
}

export const EmailCreateModal: React.FC<CreateEmailModalProps> = ({
  scenarioId,
  navigationConfig,
  directory = EmailDirectory.Inbox,
  isIntroductionEmail = false,
  routeToEmailAfterCreation = false
}) => {
  const {t} = useLucaTranslation()
  const {
    formMethods,
    createEmail,
    dismissModal,
    isConfirmDisabled,
    directoryOptions,
    senderOrRecipientInputLabelKey
  } = useEmailCreateModal(scenarioId, navigationConfig, isIntroductionEmail, routeToEmailAfterCreation, directory)
  const {register, handleSubmit, errors, control} = formMethods

  const directorySelect = (
    <Controller
      render={({onChange, value}) => (
        <CustomSelect
          customStyles={styles.selectWrapper}
          name="directory"
          onChange={onChange}
          disabled={isIntroductionEmail}
          value={value}
          optionList={directoryOptions}
          labelKey={"email__directory"}
        />
      )}
      control={control}
      rules={{required: true}}
      name="directory"
    />
  )

  const emailInput = (
    <TextInput
      ref={register({
        required: true,
        pattern: {
          value: emailRegexPattern,
          message: t("validation__email_format")
        }
      })}
      name="senderOrRecipient"
      hasValidationError={!!errors?.senderOrRecipient}
      type={InputType.email}
      labelKey={senderOrRecipientInputLabelKey}
      placeholder={t("email__create_placeholder")}
    />
  )

  return (
    <Overlay>
      <Modal
        customStyles={styles.modal}
        confirmButtonKey="create_button"
        dismissButtonKey="cancel"
        title={t("email__create_email")}
        onDismiss={dismissModal}
        onConfirm={handleSubmit(createEmail)}
        confirmButtonDisabled={isConfirmDisabled}>
        <div>
          <Text>{t("email__create_hint")}</Text>
          <div css={[styles.row, styles.grid]}>
            <div css={styles.inputWrapper}>{emailInput}</div>
            {directorySelect}
          </div>
        </div>
      </Modal>
    </Overlay>
  )
}

const styles = {
  modal: css({
    width: "50vw"
  }),
  row: css({
    marginTop: spacingLarge
  }),
  grid: css({
    display: "grid",
    gridTemplateColumns: "1fr 1fr",
    gridGap: spacingMedium
  }),
  selectWrapper: css({
    width: "100%"
  }),
  inputWrapper: css({
    "> div": css(Flex.column),

    input: {
      width: "100%"
    }
  })
}
