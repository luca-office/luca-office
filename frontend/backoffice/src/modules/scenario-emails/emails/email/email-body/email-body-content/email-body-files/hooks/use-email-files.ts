import React from "react"
import {FileUsageType, Relevance, TextDocumentCreation} from "shared/graphql/generated/globalTypes"
import {useCreateFile, useDeleteFileFromScenario, useFilesForScenario} from "shared/graphql/hooks"
import {useCreateTextDocument} from "shared/graphql/hooks/mutations/text-document/use-create-textdocument"
import {DeleteEntityHook, Email, File as AppFile, UploadBinary} from "shared/models"

export interface UseEmailFilesHook {
  readonly files: AppFile[]
  readonly filesLoading: boolean
  readonly uploadVisible: boolean
  readonly deleteFileHook: () => DeleteEntityHook
  readonly uploadBinaries: (binaries: UploadBinary[]) => void
  readonly showUpload: () => void
  readonly hideUpload: () => void
  readonly createTextDocumentFile: (textDocument: TextDocumentCreation, title: string) => void
}

export const useEmailFiles = (email: Email): UseEmailFilesHook => {
  const {id, scenarioId} = email
  const [showUpload, setShowUpload] = React.useState(false)
  const {createFile} = useCreateFile({scenarioId})
  const {files: scenarioFiles, filesLoading} = useFilesForScenario(scenarioId)
  const deleteFileFromScenario = useDeleteFileFromScenario(scenarioId)
  const files = scenarioFiles
    .map(file => file.filter(f => f.usageType === FileUsageType.Email && f.emailId === email.id))
    .getOrElse([])
  const {createTextDocument} = useCreateTextDocument({scenarioId})

  const uploadBinaries = (binaries: UploadBinary[]) => {
    Promise.all(
      binaries.map(binary =>
        createFile({
          name: binary.data.filename,
          relevance: Relevance.PotentiallyHelpful,
          tags: [],
          emailId: id,
          usageType: FileUsageType.Email,
          ...{[binary.isSpreadsheet ? "spreadsheetId" : "binaryFileId"]: binary.data.id}
        })
      )
    )
      .catch(err => {
        console.error(err)
      })
      .finally(() => setShowUpload(false))
  }

  const createTextDocumentFile = (textDocument: TextDocumentCreation, title: string) => {
    createTextDocument(textDocument).then(textDocumentOption =>
      textDocumentOption.forEach(textDocument =>
        createFile({
          name: title,
          relevance: Relevance.PotentiallyHelpful,
          tags: [],
          usageType: FileUsageType.Email,
          textDocumentId: textDocument.id,
          emailId: email.id
        })
      )
    )
  }

  return {
    hideUpload: () => setShowUpload(false),
    showUpload: () => setShowUpload(true),
    uploadVisible: showUpload,
    deleteFileHook: () => deleteFileFromScenario,
    files,
    filesLoading,
    uploadBinaries,
    createTextDocumentFile
  }
}
