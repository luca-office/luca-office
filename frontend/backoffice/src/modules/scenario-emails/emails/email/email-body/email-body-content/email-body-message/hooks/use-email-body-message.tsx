import * as React from "react"
import {EmailMessageTemplate} from "shared/enums"
import {EmailUpdate} from "shared/graphql/generated/globalTypes"
import {useSampleCompanyLazy} from "shared/graphql/hooks"
import {Email, SampleCompany, Scenario} from "shared/models"
import {useLucaTranslation} from "shared/translations"
import {
  getCaretPosition,
  getEmailMessageTemplateValue,
  getEmailMessageWithTemplateLabels,
  Option,
  setCaretPosition
} from "shared/utils"
import {useEmailUpdate} from "../../../../hooks/use-email-update"

export interface useEmailBodyMessageHook {
  readonly dataLoading: boolean
  readonly sampleCompanyId: UUID | null
  readonly sampleCompany: Option<SampleCompany>
  readonly sampleCompanyName: string | null
  readonly sampleCompanyEmailSignature: string | null
  readonly emailMessage: string
  readonly updateEmailMessage: (value: string) => void
  readonly updateEmail: (update: EmailUpdate) => void
  readonly handleSubjectUpdate: (subject: string) => void
  readonly emailMessageHasSignature: boolean
  readonly emailMessageWithTemplateLabels: string
  readonly updateEmailMessageWithTemplate: (template: EmailMessageTemplate) => void
  readonly textAreaRef: React.MutableRefObject<HTMLTextAreaElement | null>
}

export const useEmailBodyMessage = (email: Email, scenario: Option<Scenario>): useEmailBodyMessageHook => {
  const {t} = useLucaTranslation()

  const textAreaRef = React.useRef<HTMLTextAreaElement | null>(null)

  const [emailMessage, updateEmailMessage] = React.useState(email.message)

  const sampleCompanyId = scenario.map(scenario => scenario.sampleCompanyId).orNull()

  const {sampleCompany, getSampleCompany} = useSampleCompanyLazy()
  const {updateEmail, updateEmailLoading} = useEmailUpdate(email)

  const emailMessageWithTemplateLabels = getEmailMessageWithTemplateLabels(t, email.message)

  const sampleCompanyName = sampleCompany.map(company => company.name).orNull()
  const sampleCompanyEmailSignature = sampleCompany.map(company => company.emailSignature).orNull()
  const emailMessageHasSignature =
    !!sampleCompanyEmailSignature && emailMessage.indexOf(sampleCompanyEmailSignature) !== -1

  const handleSubjectUpdate = (subject: string) =>
    new Promise<void>(resolve => updateEmail({subject: subject.trim()}, resolve))

  const updateEmailMessageWithTemplate = (template: EmailMessageTemplate) => {
    const templateValue = getEmailMessageTemplateValue(template)

    if (textAreaRef.current === null) {
      updateEmailMessage(emailMessage === "" ? templateValue : `${emailMessage} ${templateValue}`)
      return
    }

    const {start, end} = getCaretPosition(textAreaRef.current)

    updateEmailMessage(
      `${emailMessage.substring(0, start)}${templateValue}${emailMessage.substring(end, emailMessage.length)}`
    )

    const caretPosition = end + templateValue.length

    // use set timeout 0 to make sure the new value was already rendered
    setTimeout(
      () =>
        textAreaRef.current !== null &&
        setCaretPosition(textAreaRef.current, {start: caretPosition, end: caretPosition}),
      0
    )
  }

  React.useEffect(() => {
    if (sampleCompanyId !== null) {
      getSampleCompany(sampleCompanyId)
    }
  }, [sampleCompanyId])

  React.useEffect(() => {
    updateEmailMessage(email.message)
  }, [email.message])

  return {
    dataLoading: updateEmailLoading,
    sampleCompanyId,
    sampleCompany,
    sampleCompanyName,
    sampleCompanyEmailSignature,
    emailMessage,
    updateEmailMessage,
    updateEmail,
    handleSubjectUpdate,
    emailMessageHasSignature,
    emailMessageWithTemplateLabels,
    updateEmailMessageWithTemplate,
    textAreaRef
  }
}
