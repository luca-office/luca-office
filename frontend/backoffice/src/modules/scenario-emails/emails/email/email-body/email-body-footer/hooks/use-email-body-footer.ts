import {useScenarioQuestionnaires} from "shared/graphql/hooks"
import {ScenarioQuestionnaire, SelectOption} from "shared/models"
import {useLucaTranslation} from "shared/translations"
import {getRelevanceCriteria} from "shared/utils"

export interface UseEmailBodyFooterHook {
  readonly relevanceOptions: SelectOption[]
  readonly scenarioQuestionnaires: ScenarioQuestionnaire[]
}

export const useEmailBodyFooter = (scenarioId: UUID): UseEmailBodyFooterHook => {
  const {t} = useLucaTranslation()

  const relevanceOptions = getRelevanceCriteria(t)

  const {scenarioQuestionnaires} = useScenarioQuestionnaires(scenarioId)

  return {
    relevanceOptions,
    scenarioQuestionnaires: scenarioQuestionnaires
      .getOrElse([])
      .filter(questionnaire => questionnaire.scenarioId === scenarioId)
  }
}
