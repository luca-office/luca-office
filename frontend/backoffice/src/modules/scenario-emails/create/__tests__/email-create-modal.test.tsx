import {act} from "@testing-library/react"
import {shallow} from "enzyme"
import * as React from "react"
import {Controller} from "react-hook-form"
import {create} from "react-test-renderer"
import {CustomSelect, Modal, Overlay, TextInput} from "shared/components"
import {scenariosMock} from "shared/graphql/__mocks__"
import {EmailDirectory} from "shared/graphql/generated/globalTypes"
import {mockedFormMethods} from "sharedTests/utils/form-methods-mock"
import wait from "waait"
import {Route} from "../../../../routes"
import {CreateEmailModalProps, EmailCreateModal} from "../email-create-modal"
import * as useEmailCreateModalHook from "../hooks/use-email-create-modal"
import {UseEmailCreateModalHook} from "../hooks/use-email-create-modal"

const scenarioId = scenariosMock[0].id
const directoryOptions = [
  {value: EmailDirectory.Inbox, label: "email__directory_inbox"},
  {value: EmailDirectory.Sent, label: "email__directory_sent"},
  {value: EmailDirectory.Trash, label: "email__directory_trash"}
]

const defaultProps: CreateEmailModalProps = {
  scenarioId,
  directory: EmailDirectory.Inbox,
  navigationConfig: {
    route: Route.ScenarioEmails,
    payload: {scenarioId, directory: EmailDirectory.Inbox}
  },
  isIntroductionEmail: false
}

const hookValuesDefault: UseEmailCreateModalHook = {
  formMethods: mockedFormMethods,
  createEmail: jest.fn(),
  createEmailLoading: false,
  dismissModal: jest.fn(),
  directoryOptions,
  senderOrRecipientInputLabelKey: "email__directory_sender",
  isConfirmDisabled: false
}

const stateSpy = jest.spyOn(useEmailCreateModalHook, "useEmailCreateModal")

describe("email-create-modal", () => {
  it("renders correctly", async () => {
    stateSpy.mockReturnValue(hookValuesDefault)
    const component = create(<EmailCreateModal {...defaultProps} />)
    await act(() => wait(0))
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("has correct structure", async () => {
    stateSpy.mockReturnValue(hookValuesDefault)
    const tree = shallow(<EmailCreateModal {...defaultProps} />)

    await act(() => wait(0))

    const overlay = tree.find(Overlay)
    expect(overlay).toHaveLength(1)

    const modal = overlay.dive().find(Modal)
    expect(modal).toHaveLength(1)

    const modalContent = modal.dive()
    expect(modalContent.find(Controller)).toHaveLength(1)
    expect(modalContent.find(Controller).dive().find(CustomSelect)).toHaveLength(1)
    expect(modalContent.find(TextInput)).toHaveLength(1)
  })
  it("has correct structure for introduction mail", async () => {
    stateSpy.mockReturnValue(hookValuesDefault)
    const tree = shallow(<EmailCreateModal {...defaultProps} isIntroductionEmail={true} />)

    await act(() => wait(0))

    const overlay = tree.find(Overlay)
    expect(overlay).toHaveLength(1)

    const modal = overlay.dive().find(Modal)
    expect(modal).toHaveLength(1)

    const modalContent = modal.dive()
    expect(modalContent.find(Controller)).toHaveLength(1)
    expect(modalContent.find(Controller).dive().find(CustomSelect)).toHaveLength(1)
    expect(modalContent.find(Controller).dive().find(CustomSelect).prop("disabled")).toBeTruthy()
    expect(modalContent.find(TextInput)).toHaveLength(1)
  })
})
