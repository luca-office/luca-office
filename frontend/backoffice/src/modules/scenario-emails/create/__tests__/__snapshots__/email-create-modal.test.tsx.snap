// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`email-create-modal renders correctly 1`] = `
<div
  className="css-1iw6hgd-Overlay"
>
  <div
    className="css-1iw6hgd-Overlay"
  >
    <div>
      <div
        className="css-1yjl9p0-Modal"
      >
        <header
          className="css-14sl779-Modal"
        >
          <div
            className="css-1f6p2vj-Heading"
          >
            email__create_email
          </div>
          <div
            className="css-12vbzj8-Icon"
            dangerouslySetInnerHTML={
              {
                "__html": {},
              }
            }
            onClick={[MockFunction]}
          />
        </header>
        <section
          className="css-1gs9x9w-Modal"
        >
          <div>
            <div
              className="css-fmthw-Text"
            >
              email__create_hint
            </div>
            <div
              className="css-8gdngw-EmailCreateModal"
            >
              <div
                className="css-zdb51k-EmailCreateModal"
              >
                <div
                  className="css-17as3x0"
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                >
                  <div
                    className="css-13s2sfk-Heading"
                  >
                    email__directory_sender
                  </div>
                  <div
                    className="css-1unfqpo"
                  >
                    <input
                      className="css-pkbj54"
                      defaultValue=""
                      disabled={false}
                      name="senderOrRecipient"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      placeholder="email__create_placeholder"
                      type="email"
                    />
                  </div>
                </div>
              </div>
              <div
                className="css-557pd0"
                data-testid="custom-select"
              >
                <div
                  className="css-0"
                >
                  <div
                    className="css-13s2sfk-Heading"
                  >
                    email__directory
                  </div>
                </div>
                <div
                  className="css-18vooac"
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                >
                  <div
                    className="custom-select css-b62m3t-container"
                    onKeyDown={[Function]}
                  >
                    <span
                      className="css-1f43avz-a11yText-A11yText"
                      id="react-select-2-live-region"
                    />
                    <span
                      aria-atomic="false"
                      aria-live="polite"
                      aria-relevant="additions text"
                      className="css-1f43avz-a11yText-A11yText"
                    />
                    <div
                      className=" css-1l51fcv-control"
                      onMouseDown={[Function]}
                      onTouchEnd={[Function]}
                    >
                      <div
                        className=" css-1fdsijx-ValueContainer"
                      >
                        <div
                          className=" css-1jqq78o-placeholder"
                          id="react-select-2-placeholder"
                        >
                          select__placeholder
                        </div>
                        <input
                          aria-autocomplete="list"
                          aria-describedby="react-select-2-placeholder"
                          aria-expanded={false}
                          aria-haspopup={true}
                          aria-readonly={true}
                          className="css-mohuvp-dummyInput-DummyInput"
                          disabled={false}
                          id="react-select-2-input"
                          inputMode="none"
                          onBlur={[Function]}
                          onChange={[Function]}
                          onFocus={[Function]}
                          role="combobox"
                          tabIndex={0}
                          value=""
                        />
                      </div>
                      <div
                        className=" css-1hb7zxy-IndicatorsContainer"
                      >
                        <span
                          className=" css-1uei4ir-indicatorSeparator"
                        />
                        <div
                          aria-hidden="true"
                          className=" css-1xc3v61-indicatorContainer"
                          onMouseDown={[Function]}
                          onTouchEnd={[Function]}
                        >
                          <div
                            className="css-l0jg45-Icon"
                            dangerouslySetInnerHTML={
                              {
                                "__html": {},
                              }
                            }
                          />
                        </div>
                      </div>
                    </div>
                    <input
                      name="directory"
                      type="hidden"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </section>
        <footer
          className="css-ukdqb3-Modal"
        >
          <button
            className="cancel-button css-a1i4d5-Button"
            disabled={false}
            onClick={[Function]}
          >
            <div
              className="css-95j2hm-Button"
            >
              cancel
            </div>
          </button>
          <button
            className="confirm-button css-1ex7tyh-Button"
            disabled={false}
            onClick={[Function]}
          >
            <div
              className="css-95j2hm-Button"
            >
              create_button
            </div>
          </button>
        </footer>
      </div>
    </div>
  </div>
</div>
`;
